hips/hip-9999.md

```diff
@@ -0,0 +1,76 @@
+---
+hip: 9999
+title: "Document backwards-compatibility rules"
+authors: [ "Marc Khouzam <marc.khouzam@montreal.ca>" ]
+created: "2020-09-18"
+type: "process"
+status: "draft"
+---
+
+## Abstract
+
+Define and document the backwards-compatibility rules that apply to Helm minor and patch releases.
+
+## Motivation
+
+Helm aims to strictly respect [Semantic Versioning][semver], where minor and patch releases must be 100% backwards-compatible.
+Backwards-compatibility implies that when moving to a new backwards-compatible release:
+* tools (such as Continuous Integration tools) using the helm CLI will not require any changes
+* programs using the Helm Go API will continue to compile 
+
+Some contributors to the Helm code base may not be comfortable with the details of what constitutes a compatibility-breaking change.
+Furthermore, in some cases, there can be a difference of opinion, even amongst the maintainers, on what constitutes such a change.
+
+Having a well-defined, documented set of backwards-compatibility rules for the Helm code-base will allow to:
+* avoid compatibility-breaking changes being merged by mistake
+* guide contributors in understanding helm's backwards-compatibility requirements
+* remove any uncertainty when it comes to enforcing backwards compatibility
+* provide a basis for discussion and possible updates to how helm handles backwards compatibility
+
+## Rationale
+
+The maintainers have agreed that a formal definition of Helm's backwards-compatibility rules would be beneficial.
+
+## Specification
+
+The following compatibility rules will apply to helm minor and patch releases:
+* exported Go APIs will remain compatible as per this [article][go-module-comp]
+* CLI commands:
+    * are case-sensitive
+    * must not be removed
+    * must not be renamed
+    * must not be moved elsewhere in the command hierarchy
+* CLI flags:
+    * are case-sensitive
+    * must not be removed
+    * must not be renamed
+    * must not be moved in the command hierarchy
+    * must not change type
```

Comment:
I assume this means "if they take an argument, the argument's type must not become more restrictive"?

e.g. `--foo=123`.

If the initial version of `--foo` takes an `int8`, can a subesequent PR:
- bump it to an int32
- parse it as a float
- treat it as the string `"132"`

The first one seems to be a yes. The second one... I am not sure. The third one is a real grey area.

Reply:
I hadn't thought of it that far 😄 But you are right that some cases can allow for type changes.
And I think your three examples are ok; well, I haven't quickly found a counter example.

How about saying:
"A flag must not change type.  Exceptions are when the new type is a superset of the original type (e.g., int8 to int32, int to float, float to string)"

What do you think?